File: 2020/23/Haskell/src/Part2.hs
module Part2 where

import           Data.List
import qualified Data.Vector.Unboxed.Mutable   as MV
import           Data.Vector.Unboxed.Mutable    ( IOVector )

type Game = IOVector Int

maxCups :: Int
maxCups = 1000000

iterations :: Int
iterations = 10000000

readInput :: String -> [Int]
readInput = map (\x -> read [x])

initInput :: [Int] -> [(Int, Int)]
initInput [_         ] = []
initInput (x : y : zs) = (x, y) : initInput (y : zs)
initInput []           = []

initCups :: (Int, Int, Int) -> [(Int, Int)] -> [Int]
initCups (_, b, m) =
  map snd . sortBy (\(x, _) (y, _) -> compare x y) . ((b, m) :)

initGame :: Int -> [Int] -> Game -> IO ()
initGame i [] g | i == maxCups - 1 = return ()
                | otherwise = MV.write g i (i + 2) >> initGame (i + 1) [] g
initGame i (x : xs) g | i == maxCups - 1 = return ()
                      | otherwise = MV.write g i x >> initGame (i + 1) xs g

stepGame :: Int -> Int -> Game -> IO ()
stepGame i n0 g
  | i == iterations = return ()
  | otherwise = do
    p1 <- MV.read g (n0 - 1)
    p2 <- MV.read g (p1 - 1)
    p3 <- MV.read g (p2 - 1)
    let ps = [p1, p2, p3]
    n1 <- MV.read g (p3 - 1)
    MV.write g (n0 - 1) n1
    let c | (n0 - 1) `notElem` ps = n0 - 1
          | (n0 - 2) `notElem` ps = n0 - 2
          | (n0 - 3) `notElem` ps = n0 - 3
          | otherwise             = n0 - 4
    let d0 | c < 1     = maxCups
           | otherwise = c
    d1 <- MV.read g (d0 - 1)
    MV.write g (d0 - 1) p1
    MV.write g (p3 - 1) d1
    stepGame (i + 1) n1 g

solve :: String -> IO (Int, Int, Int)
solve x = do
  let parsed      = readInput x
  let parseInit   = initInput parsed
  let a@(n, _, _) = (head parsed, last parsed, maximum parsed + 1)
  let gameInit    = initCups a parseInit
  gameVec <- MV.new maxCups
  initGame 0 gameInit gameVec
  MV.write gameVec (maxCups - 1) n
  stepGame 0 n gameVec
  r1 <- MV.read gameVec 0
  r2 <- MV.read gameVec (r1 - 1)
  return (r1, r2, r1 * r2)

File: 2020/23/Rust/src/part_2.rs
use std::convert::TryFrom;

const MAX_CUPS: usize = 1_000_000;
const ITERATIONS: usize = 10_000_000;

fn i32_to_usize(x: i32) -> usize {
    usize::try_from(x).unwrap()
}

fn init_input(cups: &[i32]) -> Vec<(i32, i32)> {
    (0..cups.len() - 1)
        .map(|i| (cups[i], cups[i + 1]))
        .collect()
}

fn init_cups(init: &[(i32, i32)], last: &i32, next: &i32) -> Vec<i32> {
    let mut init = init.to_owned();
    init.push((*last, *next));
    init.sort_unstable_by(|(x, _), (y, _)| x.cmp(y));
    init.iter().map(|(_, x)| x.to_owned()).collect()
}

fn init_game(init: &[i32], mut game: Vec<usize>) -> Vec<usize> {
    let mut offset = 0;
    for cup in init {
        game.push(i32_to_usize(*cup));
        offset += 1;
    }
    for i in (offset + 2)..=MAX_CUPS {
        game.push(i);
    }
    game
}

fn step_game(mut n_0: usize, mut game: Vec<usize>) -> Vec<usize> {
    for _ in 0..ITERATIONS {
        let p_1 = game[n_0 - 1];
        let p_2 = game[p_1 - 1];
        let p_3 = game[p_2 - 1];
        let ps = vec![p_1, p_2, p_3];
        let n_1 = game[p_3 - 1];
        game[n_0 - 1] = n_1;
        let c = {
            if !ps.contains(&(n_0 - 1)) {
                n_0 - 1
            } else if !ps.contains(&(n_0 - 2)) {
                n_0 - 2
            } else if !ps.contains(&(n_0 - 3)) {
                n_0 - 3
            } else {
                n_0 - 4
            }
        };
        let d_0 = {
            if c < 1 {
                MAX_CUPS
            } else {
                c
            }
        };
        let d_1 = game[d_0 - 1];
        game[d_0 - 1] = p_1;
        game[p_3 - 1] = d_1;
        n_0 = n_1;
    }
    game
}

pub fn solve(cups: &[i32]) -> (usize, usize, u64) {
    let init = init_input(cups);
    let head = cups.first().unwrap();
    let last = cups.last().unwrap();
    let next = cups.iter().max().unwrap() + 1;
    let init = init_cups(&init, last, &next);
    let mut game = Vec::with_capacity(MAX_CUPS);
    game = init_game(&init, game);
    game.push(i32_to_usize(*head));
    game = step_game(i32_to_usize(*head), game);
    let r_1 = game[0];
    let r_2 = game[r_1 - 1];
    (r_1, r_2, u64::try_from(r_1 * r_2).unwrap())
}
